### PR TITLE
fix: device-native stick-dim reconstruction in coarse_tile layout resize

### PR DESCRIPTION
**Pull request by dgrove-oss**
#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Fixes two sites in `coarse_tile.py` that re-derive the within-stick host
dimension using a fragile heuristic: matching `stride_map[-1]` against
freshly-computed contiguous host strides to guess a `dim_order`, then
rebuilding the `SpyreTensorLayout` from scratch.

**Bug 1 — misidentified stick dim for transposed layouts.**
For a `[256, 128]` buffer with `dim_order=[1, 0]` (stick on dim 0, as produced
by `model_utils.py` for transposed Linear weights), tiling the non-stick dim
gives:

| | `device_size` | `stride_map` |
|---|---|---|
| Correct | `[4, 64, 64]` | `[4096, 1, 64]` |
| Old heuristic | `[1, 256, 64]` | `[64, 64, 1]` |

**Bug 2 — silent drop of `element_arrangement`.**
The old reconstruction used the 4-arg `SpyreTensorLayout(size, strides, dtype,
dim_order)` constructor, which always defaults `element_arrangement` to
`STANDARD`. This silently corrupted EXX2/QFP8/DL16 layouts.

**Fix — device-native reconstruction via `_resize_device_layout`.**
A new helper derives the updated layout directly from the original
`SpyreTensorLayout` without guessing a `dim_order`:

1. Match non-stick device dims to host dims by `device_size[j] == old_host_size[p]`
   (with a contiguous-stride tiebreaker for ambiguous sizes).
2. Identify the stick host dim `p*` by elimination (the unmatched remainder).
3. Update `device_size` and `stride_map` for each classified dim; leave
   non-contiguous strides unchanged.
4. Copy `device_dtype` and `element_arrangement` verbatim.

Raises `RuntimeError` for layouts that cannot be classified unambiguously
rather than silently producing a wrong result. `_allocate_full_buffer` catches
this for non-standard HBM layouts (e.g. post-restickify buffers) and falls
back to a row-major allocation with a DEBUG log.

Also adds a TODO comment in `pass_utils.py:lower_pad_sequence` at the
analogous `sm_last`-matching site, flagging it as a follow-up candidate for
the same fix.

#### Which issue(s) this PR is related to:

part of #1358 

#### Special notes for your reviewer:

- `_resize_device_layout` is placed after `_stamp_group` and before
  `_divide_ranges` in the file — it is only called from those two sites.
- The `_allocate_full_buffer` fallback path (row-major for post-restickify
  layouts) preserves existing behavior for `test_hint_restickify_stays_in_group`.
  That test only verifies generated source structure (LoopSpec presence), not
  data correctness; the correctness of the fallback for that case is a
  pre-existing open question noted in `pass_utils.py` TODO.
- `pass_utils.py:lower_pad_sequence` has the same `sm_last` pattern with a
  `raise RuntimeError` on no-match (already fail-loud). It is left as a
  follow-up; the TODO comment is the only change to that file.

#### Does this PR introduce a user-facing change?

No — internal compiler pass only. Models compiled with `spyre_hint` that
involve transposed-stick buffers or non-STANDARD `element_arrangement` will now
produce correct tiled layouts; previously they would have silently compiled with
a wrong device layout.

#### Additional note:

The `_resize_device_layout` algorithm was verified by unit tests covering:

- Transposed-stick layout (`[256, 128]`, `dim_order=[1, 0]`) — headline regression
- `element_arrangement` preservation (EXX2)
- Stride-collision case (`[2, 2, 2, 16]` where a non-stick stride equals the
  tile-count stride)
- Fail-loud path (ambiguous `pstar` → `RuntimeError`)

### CODE REVIEW

**Review by ani300 — CHANGES_REQUESTED** [review]
## Review (shared from local analysis)

The headline fix works — I empirically confirmed transposed-stick layouts (2D and 3D `dim_order=[1,0]`/`[2,0,1]`) now reconstruct correctly and `element_arrangement` is preserved. But the value-based dim classifier in `_resize_device_layout` has **two silent-miscompile paths**, and the `_divide_ranges` route lost the old code's never-raise guarantee. All results below were produced against **this PR's code** (loaded from the PR branch, not the installed base package), with repro + drop-in failing tests attached at the end.

### Blockers

**1. Tile-count dim misclassified as non-stick on size collision → silently wrong layout.** `coarse_tile.py:1295`

When the stick tile-count `ceil(stick/eps)` equals a non-stick host dim's size, Pass 1 claims the tile-count device dim as a non-stick match, so Pass 3 never recomputes it. No exception is raised, so the wrong layout flows into the scatter-copy address computation (which the comment at `:668-672` says depends on both buffers agreeing on physical layout). Realistic small shapes (batch=2, heads=4):

```
[2,128] tile STICK dim1 -> [2,64]:  got ds=[2,2,64]  want ds=[1,2,64]   WRONG
[2,128] tile dim0 -> [1,128]:       got ds=[1,1,64]  want ds=[2,1,64]   WRONG
[4,256] tile dim0 -> [2,256]:       got ds=[2,2,64]  want ds=[4,2,64]   WRONG
[3,128] tile dim1 (no collision):   got ds=[1,3,64]  want ds=[1,3,64]   OK (control)
```

`test_divide_ranges_stride_collision` claims to cover this, but its `[2,2,2,16]` shape doesn't actually trigger the misclassification — the test passes while the bug is live.

**2. Grow path skips size-1 dims → silently wrong full-buffer layout.** `coarse_tile.py:1293`

`_allocate_full_buffer` calls `_resize_device_layout(per_tile, tile_size, full_size)` — the grow direction. Pass 1's `if dsz == 1: continue` skips any dim that was tiled down to extent 1, so it never grows back. It returns the wrong layout **without raising**, so the `except RuntimeError` row-major fallback never engages:

```
[1,128] grow dim0 -> [4,128]:  got ds=[2,1,64] sm=[64,-1,1]  want ds=[2,4,64] sm=[64,128,1]  WRONG
```

(Related: Pass-4 inner-stick update at `:1388` lacks the `orig_sm[j] == -1` rescue that Pass 3 has at `:1380`.)

Both stem from the same root: classifying device dims by their numeric `device_size`/`stride_map` is ambiguous, whereas the C++ layout knows the classification structurally from `dim_order`.

### Suggestions

- **Uncaught `RuntimeError` in `_divide_ranges`** (`:1495`): the old heuristic never raised; `_allocate_full_buffer` wraps the call in `try/except` but `_divide_ranges` does not, so an unclassifiable layout (including the "post-restickify HBM" class the description names as unsupported) now aborts compilation from the tiling path. Wrap it symmetrically or use a typed sentinel instead of generic `RuntimeError`.
- **Right altitude — preserve `dim_order` and delegate to C++.** The 164-line helper re-implements `get_generic_stick_layout` + `dim_map_to_stride_map` in Python *only* because `dim_order` isn't readable off a `SpyreTensorLayout` (`module.cpp:208-212` exposes `device_size`/`stride_map`/`device_dtype`/`element_arrangement` as `def_readonly`, not `dim_order`). Exposing `dim_order` (or adding a C++ `resize()`) collapses the helper to the existing 5-arg constructor that `pass_utils.py:1126` already uses — and dissolves both blockers, since C++ computes the layout authoritatively. The `pass_utils.py` TODO would then point at the constructor rather than entrenching a third copy of the math.
- **Exceptions-as-control-flow** (`:681`): a generic `RuntimeError` raised from 4 distinct sites is used as a "couldn't classify" signal; a genuine bug inside the helper would be swallowed into the row-major fallback.
- **Reduction `pstar=None` branch** (`:1320`, commit 2) is a patch on the one-unmatched-stick assumption; the layer-correct fix above makes it unnecessary.
- **Test gaps:** grow path, tiling the stick dim itself, the collision shape that actually misclassifies, rank-3/4 transposed with multiple non-stick dims, QFP8/DL16 arrangements (only EXX2 tested), and `RuntimeError` propagation through `_divide_ranges`. The attached tests cover the first three.

### Questions

1. Do real `assign_dim_hints` graphs produce a non-stick host dim equal to `ceil(stick/eps)` (Blocker 1), or a tiled dim reaching extent 1 on the grow path (Blocker 2)? Both are silent miscompiles, so even occasional occurrence is serious.
2. Is exposing `dim_order` (or a C++ `resize`) viable? It removes most of this PR's surface and both blockers.
3. Does the `_allocate_full_buffer` row-major fallback (which discards stick orientation, keeping only `element_arrangement`) actually address the scatter copy correctly for post-restickify buffers? The description calls this a pre-existing open question.

### Verified NOT bugs

Transposed reconstruction (the headline fix) is correct on stick and non-stick tiling, 2D and 3D; the 4-arg call binds to the device-native constructor (no silent host-recompute); `elems_per_stick()`/`device_dtype`/`element_arrangement` are bound/readable; `element_arrangement` is preserved.

### Verdict: request changes

The transposed fix and dtype/arrangement handling are genuine improvements, but two confirmed silent-miscompile paths, a misleading existing test, and a new hard-abort path on the tiling route block it. Fixing the blockers — ideally by preserving `dim_order` and delegating to the C++ constructor — and adding the failing tests below would get this in.

---

<details>
<summary><b>Repro script</b> — <code>repro_pr2912.py</code> (prints got/want for both bugs + control + sanity)</summary>

Run with `TORCH_DEVICE_BACKEND_AUTOLOAD=0 python repro_pr2912.py` from a checkout of this PR with the C++ extension built. Adjust `PR_SRC` to your checkout path. It loads `_resize_device_layout` from source so it exercises the PR's code, not the installed package.

```python
import re
import torch
from torch._inductor.ir import FlexibleLayout
from torch_spyre._C import SpyreTensorLayout

PR_SRC = "torch_spyre/_inductor/coarse_tile.py"  # path to this PR's coarse_tile.py


def load_pr_function():
    text = open(PR_SRC).read()
    start = text.index("def _resize_device_layout(")
    rest = text[start:]
    end = re.search(r"\ndef _divide_ranges\(", rest).start()
    ns = {"SpyreTensorLayout": SpyreTensorLayout}
    exec("from torch._inductor.ir import FlexibleLayout\n" + rest[:end], ns)
    return ns["_resize_device_layout"]


_resize = load_pr_function()


def ground_truth(host, dim_order):
    strides = [int(s) for s in FlexibleLayout.contiguous_strides(host)]
    return SpyreTensorLayout(host, strides, torch.float16, dim_order)


def case(host_old, dim_order, host_new, label):
    so = [int(s) for s in FlexibleLayout.contiguous_strides(host_old)]
    stl = SpyreTensorLayout(host_old, so, torch.float16, dim_order)
    gt = ground_truth(host_new, dim_order)
    try:
        res = _resize(stl, host_old, host_new)
    except Exception as e:
        print(f"  {label}: RAISED {type(e).__name__}")
        return False
    got = (list(res.device_size), list(res.stride_map))
    want = (list(gt.device_size), list(gt.stride_map))
    ok = got == want
    print(f"  {label}")
    print(f"     orig  ds={list(stl.device_size)} sm={list(stl.stride_map)} eps={stl.elems_per_stick()}")
    print(f"     got   ds={got[0]} sm={got[1]}")
    print(f"     want  ds={want[0]} sm={want[1]}   -> {'OK' if ok else 'WRONG (silent miscompile)'}")
    return ok


print("BUG 1: stick tile-count value collides with a non-stick host dim size")
case([2, 128], [0, 1], [2, 64], "[2,128] tile STICK dim1 -> [2,64]")
case([2, 128], [0, 1], [1, 128], "[2,128] tile non-stick dim0 -> [1,128]")
case([4, 256], [0, 1], [2, 256], "[4,256] tile non-stick dim0 -> [2,256]")
case([3, 128], [0, 1], [3, 64], "[3,128] tile STICK dim1 -> [3,64]  (control)")

print("BUG 2: grow path skips size-1 dims")
case([1, 128], [0, 1], [4, 128], "[1,128] grow non-stick dim0 -> [4,128]")

print("SANITY: cases the PR fixes correctly (must stay OK)")
case([256, 128], [1, 0], [256, 64], "transposed tile non-stick -> [256,64]")
case([256, 128], [1, 0], [128, 128], "transposed tile STICK -> [128,128]")
case([8, 256, 128], [2, 0, 1], [8, 256, 64], "3D transposed tile dim2 -> [8,256,64]")
```

Output against this PR: BUG 1 (all three) and BUG 2 print `WRONG`; the control and all three SANITY cases print `OK`.

</details>

<details>
<summary><b>Drop-in failing tests</b> — paste into <code>TestDivideRanges</code> in <code>tests/inductor/test_coarse_tiling.py</code></summary>

Against this PR: 3 fail (the bugs), 1 passes (the no-collision control). They should pass once dims are classified by `dim_order` rather than by value-matching `device_size`.

```python
def test_tilecount_collides_with_nonstick_size_tile_stick(self):
    """BUG 1: when ceil(stick/eps) equals a non-stick host dim's size, the stick
    tile-count device dim is misclassified as a non-stick match and is never
    recomputed when the stick dim is tiled."""
    from torch._inductor.ir import FlexibleLayout
    from torch_spyre._C import SpyreTensorLayout
    from torch_spyre._inductor.coarse_tile import _resize_device_layout

    def stl(host):
        s = [int(x) for x in FlexibleLayout.contiguous_strides(host)]
        return SpyreTensorLayout(host, s, torch.float16, [0, 1])

    got = _resize_device_layout(stl([2, 128]), [2, 128], [2, 64])
    want = stl([2, 64])
    self.assertEqual(list(got.device_size), list(want.device_size))
    self.assertEqual(list(got.stride_map), list(want.stride_map))


def test_tilecount_collides_with_nonstick_size_tile_nonstick(self):
    """BUG 1, non-stick variant: tiling the non-stick dim under the same
    collision shrinks the wrong dim and freezes the tile-count."""
    from torch._inductor.ir import FlexibleLayout
    from torch_spyre._C import SpyreTensorLayout
    from torch_spyre._inductor.coarse_tile import _resize_device_layout

    def stl(host):
        s = [int(x) for x in FlexibleLayout.contiguous_strides(host)]
        return SpyreTensorLayout(host, s, torch.float16, [0, 1])

    got = _resize_device_layout(stl([2, 128]), [2, 128], [1, 128])
    want = stl([1, 128])
    self.assertEqual(list(got.device_size), list(want.device_size))
    self.assertEqual(list(got.stride_map), list(want.stride_map))


def test_grow_path_size_one_dim(self):
    """BUG 2: the _allocate_full_buffer grow direction (tile_size -> full_size)
    skips per-tile dims of extent 1, so they never grow back."""
    from torch._inductor.ir import FlexibleLayout
    from torch_spyre._C import SpyreTensorLayout
    from torch_spyre._inductor.coarse_tile import _resize_device_layout

    def stl(host):
        s = [int(x) for x in FlexibleLayout.contiguous_strides(host)]
        return SpyreTensorLayout(host, s, torch.float16, [0, 1])

    got = _resize_device_layout(stl([1, 128]), [1, 128], [4, 128])
    want = stl([4, 128])
    self.assertEqual(list(got.device_size), list(want.device_size))
    self.assertEqual(list(got.stride_map), list(want.stride_map))


def test_no_collision_control_still_passes(self):
    """Control: with no size collision the existing logic is correct."""
    from torch._inductor.ir import FlexibleLayout
    from torch_spyre._C import SpyreTensorLayout
    from torch_spyre._inductor.coarse_tile import _resize_device_layout

    def stl(host):
        s = [int(x) for x in FlexibleLayout.contiguous_strides(host)]
        return SpyreTensorLayout(host, s, torch.float16, [0, 1])

    got = _resize_device_layout(stl([3, 128]), [3, 128], [3, 64])
    want = stl([3, 64])
    self.assertEqual(list(got.device_size), list(want.device_size))
    self.assertEqual(list(got.stride_map), list(want.stride_map))
```

</details>

<sub>Review analysis assisted by Claude Code; all findings reproduced empirically against the PR branch.</sub>

**Comment by dgrove-oss** [comment]
@tardieu and I had a long discussion about possible approaches to tackling this bug.   

It is worth documenting a key architectural decision that was reached.
 
We made an intentional decision to accept additional code complexity to avoid storing the `dim_order` (or equivalent) in `FixedTiledLayout`.   The combination of size, stride, stride_map, and device_size is always enough.

**Comment by dgrove-oss** [comment]
Both blockers fixed in commit 5cbee8b.

**Blocker 1 (tile-count/size collision):** The root cause was that Pass 1
accepted a size-only match even when `stride_map[j] != contiguous_stride[p]`.
The fix (Pass 1b) runs after pstar is identified: any provisional match where
`orig_sm[j] != old_hs[p]` and `orig_ds[j] == ceil(old_host_size[pstar]/eps)`
is re-classified as a tile-count dim and moved to `unmatched_j`. This preserves
correct handling of non-contiguous (permuted) non-stick dims, which also have a
stride mismatch but do *not* have the tile-count size.

**Blocker 2 (grow path / size-1 skipped):** `device_size==1` dims are now
matched by size to host dims of size 1. Sparse placeholders (no size-1 host
counterpart) are silently skipped. Pass 2 adds the `orig_sm[j] == -1` rescue
(matching the existing rescue in Passes 3 and 4) so growing from a size-1
singleton correctly updates the stride.

**Pass 4 inner-stick `orig_sm[j] == -1` rescue:** Added.

**Suggestion 2 (expose `dim_order` from C++):** `dim_order` is not stored on
the C++ struct — consumed at construction and discarded.  We had a long discussion on this,
and do not want to go down this path. 

**Suggestion 3 (exceptions as control flow):** Agreed a narrower exception type
would be cleaner; deferred as a follow-up cleanup.

**Three questions:**
1. The tile-count/size collision can occur whenever a non-stick dim happens to
   equal `ceil(stick/eps)` — e.g. any [2, 128] or [4, 256] input. The grow
   path is hit whenever a per-tile buffer has a size-1 dim from coarse tiling.
   Both are plausible in real graphs.
2. Addressed above — C++ change not desired.
3. The `_allocate_full_buffer` row-major fallback for post-restickify buffers:
   the existing test only verifies source structure, not data correctness.
   Whether the scatter copy is correct there depends on `work_division.py`'s
   address computation — called out as an open question in the PR description.

**Review by ani300 — APPROVED** [review]
lgtm after fixes, thanks!

**Review by avery-blanchard — APPROVED** [review]
looks good to me
